File: io/DataIO.java
package common.io;

import java.util.function.BiFunction;
import common.pack.Context;

public strictfp abstract class DataIO {

	/**
	 * write a number n into a byte[] start from index
	 */

	public static void fromByte(byte[] b, int index, byte n) {
		b[index] = n;
	}

	public static void fromDouble(byte[] b, int index, double n) {
		long l = Double.doubleToLongBits(n);
		fromLong(b, index, l);
	}

	public static void fromFloat(byte[] b, int index, float n) {
		int l = Float.floatToIntBits(n);
		fromInt(b, index, l);
	}

	public static void fromInt(byte[] b, int index, int n) {
		for (int i = 0; i < 4; i++)
			b[index + i] = (byte) (n >> (i * 8) & 0xff);
	}

	public static void fromLong(byte[] b, int index, long n) {
		for (int i = 0; i < 8; i++)
			b[index + i] = (byte) (n >> (i * 8) & 0xff);
	}

	public static void fromShort(byte[] b, int index, short n) {
		for (int i = 0; i < 2; i++)
			b[index + i] = (byte) (n >> (i * 8) & 0xff);
	}

	public static byte[] getSignature(int n) {
		byte[] ans = new byte[4];
		fromInt(ans, 0, n);
		return ans;
	}

	public static <T> T readData(Context.SupExc<Integer> con, int n, BiFunction<int[], Integer, T> func)
			throws Exception {
		int[] data = new int[n];
		for (int i = 0; i < n; i++)
			data[i] = con.get();
		return func.apply(data, 0);
	}

	public static int readInt(Context.SupExc<Integer> con) throws Exception {
		int[] data = new int[4];
		for (int i = 0; i < 4; i++)
			data[i] = con.get();
		return toInt(data, 0);
	}

	/**
	 * read a number from a byte[] start from index
	 */

	public static int toByte(int[] datas, int index) {
		return datas[index];
	}

	public static double toDouble(int[] datas, int index) {
		return Double.longBitsToDouble(toLong(datas, index));
	}

	public static float toFloat(int[] datas, int index) {
		return Float.intBitsToFloat(toInt(datas, index));
	}

	public static int toInt(byte[] datas, int index) {
		int ans = 0;
		for (int i = 0; i < 4; i++)
			ans += (datas[index + i] & 0xff) << (i * 8);
		return ans;
	}

	public static int toInt(int[] datas, int index) {
		int ans = 0;
		for (int i = 0; i < 4; i++)
			ans += (datas[index + i]) << (i * 8);
		return ans;
	}

	public static long toLong(int[] datas, int index) {
		long l = 0;
		for (int i = 0; i < 8; i++) {
			long temp = datas[index + i];
			temp <<= i * 8;
			l += temp;
		}
		return l;
	}

	public static int toShort(int[] datas, int index) {
		short ans = 0;
		for (int i = 0; i < 2; i++)
			ans += (datas[index + i]) << (i * 8);
		return ans;
	}

	/**
	 * translate byte[] to int[] in order to make all the bytes unsigned
	 */

	public static int[] translate(byte[] datas) {
		int[] ans = new int[datas.length];
		for (int i = 0; i < ans.length; i++)
			ans[i] = (datas[i]) & 0xff;
		return ans;
	}

	public static byte[] translate(int[] datas) {
		byte[] ans = new byte[datas.length];
		for (int i = 0; i < ans.length; i++)
			ans[i] = (byte) datas[i];
		return ans;
	}

}

File: io/DataIO.kt
package common.io

import common.system.P

@Strictfp
abstract class DataIO {
    companion object {
        protected const val BYTE: Byte = 1
        protected const val SHORT: Byte = 2
        protected const val INT: Byte = 3
        protected const val LONG: Byte = 4
        protected const val FLOAT: Byte = 5
        protected const val DOUBLE: Byte = 6
        protected const val BYTESI: Byte = 7
        protected const val INTSB: Byte = 8
        protected const val INTSI: Byte = 9
        protected const val STRING: Byte = 10
        protected const val BYTESB: Byte = 11
        protected const val INTSSBB: Byte = 12
        protected const val DOUBLESB: Byte = 13
        protected const val SUBS: Byte = 14
        fun fromABP(b: ByteArray, index: Int, n: P) {
            var index = index
            fromShort(b, index, (n.x * 10 + 2000) as Short)
            index += 2
            fromShort(b, index, (n.y * 10 + 2000) as Short)
        }

        /** write a number n into a byte[] start from index  */
        fun fromByte(b: ByteArray, index: Int, n: Byte) {
            b[index] = n
        }

        fun fromDouble(b: ByteArray, index: Int, n: Double) {
            val l = java.lang.Double.doubleToLongBits(n)
            fromLong(b, index, l)
        }

        fun fromFloat(b: ByteArray, index: Int, n: Float) {
            val l = java.lang.Float.floatToIntBits(n)
            fromInt(b, index, l)
        }

        fun fromInt(b: ByteArray, index: Int, n: Int) {
            for (i in 0..3) b[index + i] = (n shr i * 8 and 0xff).toByte()
        }

        fun fromLong(b: ByteArray, index: Int, n: Long) {
            for (i in 0..7) b[index + i] = (n shr i * 8 and 0xff).toByte()
        }

        fun fromShort(b: ByteArray, index: Int, n: Short) {
            for (i in 0..1) b[index + i] = (n shr i * 8 and 0xff) as Byte
        }

        fun getSignature(n: Int): ByteArray {
            val ans = ByteArray(4)
            fromInt(ans, 0, n)
            return ans
        }

        fun toABP(datas: IntArray, index: Int): P {
            val x = toShort(datas, index).toDouble()
            val y = toShort(datas, index + 2).toDouble()
            return P(x * 0.1 - 200, y * 0.1 - 200)
        }

        /** read a number from a byte[] start from index  */
        fun toByte(datas: IntArray, index: Int): Int {
            return datas[index]
        }

        fun toDouble(datas: IntArray, index: Int): Double {
            return java.lang.Double.longBitsToDouble(toLong(datas, index))
        }

        fun toFloat(datas: IntArray, index: Int): Float {
            return java.lang.Float.intBitsToFloat(toInt(datas, index))
        }

        fun toInt(datas: IntArray, index: Int): Int {
            var ans = 0
            for (i in 0..3) ans += datas[index + i] shl i * 8
            return ans
        }

        fun toLong(datas: IntArray, index: Int): Long {
            var l: Long = 0
            for (i in 0..7) {
                var temp = datas[index + i].toLong()
                temp = temp shl i * 8
                l += temp
            }
            return l
        }

        fun toShort(datas: IntArray, index: Int): Int {
            var ans: Short = 0
            for (i in 0..1) (ans += datas[index + i] shl (i * 8).toShort().toInt()).toShort()
            return ans.toInt()
        }

        /** translate byte[] to int[] in order to make all the bytes unsigned  */
        fun translate(datas: ByteArray): IntArray {
            val ans = IntArray(datas.size)
            for (i in ans.indices) ans[i] = datas[i] and 0xff
            return ans
        }

        fun translate(datas: IntArray): ByteArray {
            val ans = ByteArray(datas.size)
            for (i in ans.indices) ans[i] = datas[i].toByte()
            return ans
        }
    }
}
